frontend/packages/core-auth/src/components/AuthWrapper.tsx

```diff
@@ -114,16 +111,57 @@ const AuthWrapper: React.FC<AuthWrapperProps> = ({
     }
   }, [tokenResponse, isLoggedIn]);
 
+  // Schedule a token refresh before it expires (or use an existing schedule)
+  const refreshTokenTimeout = useRef<number | null>(null);
   useEffect(() => {
-    onTokenChange(token);
-  }, [token]);
+    if (
+      tokenResponse != null &&
+      tokenResponse.expiresIn != null &&
+      tokenResponse.expiresIn > 0
+    ) {
+      if (refreshTokenTimeout.current != null)
+        window.clearTimeout(refreshTokenTimeout.current);
+
+      const delay = tokenResponse.getExpiryMs() - Date.now();
+      console.log({ delay });
+
+      if (delay <= 0) {
+        refreshToken();
+      } else {
+        refreshTokenTimeout.current = window.setTimeout(
+          () => refreshToken(),
+          delay,
+        );
+      }
```

Comment:
I feel this is going to end up with some failed requests if you make a request while refreshing. It may be better to refresh more frequently while the access token is still valid. Eg. do this on a `setInterval` at 1/3 of the expiration time and always refresh.